src/main/java/reposense/authorship/analyzer/AuthorshipAnalyzer.java

```diff
@@ -0,0 +1,267 @@
+
+        // Give full credit if author is unknown, is before since date, is in ignored list, or is an ignored file
+        if (deletedLineInfo.getAuthor().equals(Author.UNKNOWN_AUTHOR)
+                || deletedLineInfo.getTimestampMilliseconds() < sinceDateInMilliseconds
+                || CommitHash.isInsideCommitList(deletedLineInfo.getCommitHash(), config.getIgnoreCommitList())
+                || deletedLineInfo.getAuthor().isIgnoringFile(Paths.get(deletedLine.getFilePath()))) {
+            return true;
+        }
+
+        // Give partial credit if currentAuthor is not the author of the previous version
+        if (!currentAuthor.equals(deletedLineInfo.getAuthor())) {
+            return false;
+        }
+
+        // Check the previous version as currentAuthor is the same as author of the previous version
+        return analyzeAuthorship(config, deletedLine.getFilePath(), deletedLine.getLineContent(),
+                deletedLineInfo.getCommitHash(), deletedLineInfo.getAuthor(), originalityThreshold);
+    }
+
+    /**
+     * Returns the deleted line in {@code commitHash} that has the lowest originality with {@code lineContent}.
+     */
+    private static CandidateLine getDeletedLineWithLowestOriginality(RepoConfiguration config, String filePath,
+            String lineContent, String commitHash) {
+        CandidateLine lowestOriginalityLine = null;
+
+        String gitLogCacheKey = config.getRepoRoot() + commitHash;
+        String[] parentCommits;
+        if (GIT_LOG_CACHE.containsKey(gitLogCacheKey)) {
+            parentCommits = GIT_LOG_CACHE.get(gitLogCacheKey);
+        } else {
+            String gitLogResults = GitLog.getParentCommits(config.getRepoRoot(), commitHash);
+            parentCommits = gitLogResults.split(" ");
+            GIT_LOG_CACHE.put(gitLogCacheKey, parentCommits);
+        }
+
+        for (String parentCommit : parentCommits) {
+            String gitDiffCacheKey = config.getRepoRoot() + parentCommit + commitHash;
+            ArrayList<String> fileDiffResultList;
+            ArrayList<String> preImageFilePathList;
+            ArrayList<String> postImageFilePathList;
+
+            if (GIT_DIFF_CACHE.containsKey(gitDiffCacheKey)) {
+                ArrayList<ArrayList<String>> cacheResults = GIT_DIFF_CACHE.get(gitDiffCacheKey);
+                fileDiffResultList = cacheResults.get(0);
+                preImageFilePathList = cacheResults.get(1);
+                postImageFilePathList = cacheResults.get(2);
+            } else {
+                fileDiffResultList = new ArrayList<>();
+                preImageFilePathList = new ArrayList<>();
+                postImageFilePathList = new ArrayList<>();
+
+                // Generate diff between commit and parent commit
+                String gitDiffResult = GitDiff.diffCommits(config.getRepoRoot(), parentCommit, commitHash);
+                String[] fileDiffResults = gitDiffResult.split(DIFF_FILE_CHUNK_SEPARATOR);
```

Comment:
Should we look into creating a `Pattern` object for `DIFF_FILE_CHUNK_SEPARATOR` before the `for` loop and use the `Matcher` object from `Pattern::matcher` to split the string?

Reply:
Just to double check, this only applies to `split` functions called within a loop? What about the case if the `split` is called in a method, but this method is called from within a loop? Because of resolving the 4 level nested method, this part of the code is now extracted out as a new method.

Reply:
Sorry for the delayed response, but if I'm not mistaken, any time where Regex operations are used directly without `Matcher` or `Pattern` causes the creation of one `Matcher` or `Pattern` object, since these operations use the 2 classes to execute the Regex task.

In the case that you have outlined, I believe we can treat it as multiple `Pattern` and `Matcher` objects being created throughout the loop.

Would it be possible to declare a `Pattern(DIFF_FILE_CHUNK_SEPARATOR)` object at the class level, and subsequently reused by the different methods for the various Regex operations within the nested function calls?

Reply:
That sounds like a great idea, perhaps we can have something in `StringUtil` to do this. We can even cache the pattern object to avoid recreating them.

This seems to be out of the scope of my project. We can make a separate issue addressing this and perhaps also check through the code base for other instances.

Reply:
Also similar to what I have mentioned above, I think in the general case for `split` operation, I'm trying to get an array of results and work on that. Using `pattern` and `matcher` will result in a for loop to iterate and using `substring` to extract out the substrings. I'm not sure if this trade off will result in better or worse runtime/memory performance, but definitely there will be more things to handle and the code will be less clean.

So I think that I will stick with what I have for now.